### ISSUE TITLE
getting ready for the Amber18 release

### ISSUE THREAD

**Issue by dacase** [Amber-MD/cmake-buildscripts#108]
I'm starting this issue, just to collect various (mostly small) tasks that we need for the Amber18 release.  I'll add things as they come to mind, and delete things that get resolved.

1. INSTALL_TOOLS needs to add a test-only Makefile to the top level directory, so that running the tests looks the same for legacy and CMake builds.  A file that I think will work is attached.  I'd do this, except am short of time and don't really understand how INSTALL_TOOLS knows what to do.
[Makefile.txt](https://github.com/Amber-MD/cmake-buildscripts/files/1789873/Makefile.txt)

...more later....

**Comment by multiplemonomials** [comment]
Personally, I'd at least like to fix #53, #37, and #14 before the release, though those are all large jobs.

**Comment by multiplemonomials** [comment]
OK I added your Makefile in Amber master `3ab511ae3d0f`.  It seems to work fine.

**Comment by dacase** [comment]
Not sure if anyone else is seeing this (MacOS, gcc 6.2), no -mpi flag sent to CMake:

Scanning dependencies of target sff_rism_interface
[ 24%] Building Fortran object AmberTools/src/rism/CMakeFiles/sff_rism_interface.dir/amber_rism_interface.F90.o
/home/case/test/amber18/AmberTools/src/rism/amber_rism_interface.F90:1452:78:

              call corr_drift(ff, rism_3d%solute%mass, rism_3d%solute%numAtoms)
                                                                              1
Error: Missing actual argument for argument 'rank' at (1)

I don't have time now to track this down, but it looks like corr_drift (or its mod file) must have been created with MPI enabled, but amber_rism_interface.F90 is being compiled in serial mode.

**Comment by multiplemonomials** [comment]
I haven't heard of anyone else having this issue, and the CI builds are
still working fine.  It might be an old module file left around by some
past build.  Can you try deleting your build folder and building again?

**Comment by dacase** [comment]
I'm having all kinds of troubles with the latest release candidate.  On Ubuntu 16.04, I tried this:

cmake /home/case/test/amber18 -DCMAKE_INSTALL_PREFIX=/home/case/test/amber_ci -DCOMPILER=GNU -DBUILD_GUI=FALSE

When it gets to building libcpptraj.so, it seems not to be linking in -libgfortran:

[ 71%] Linking CXX shared library libcpptraj.so
/home/casegroup/amber18/lib/libarpack.a(dsaupd.o): In function `dsaupd_':
dsaupd.f:(.text+0x2d7): undefined reference to `_gfortran_st_write'
dsaupd.f:(.text+0x2df): undefined reference to `_gfortran_st_write_done'

...etc...etc

We've been through this before, but I don't remember where the fix was put.  It's probably something that didn't get correctly transfered to the release branch.  But everything in $AMBERHOME/cmake matches what is currently in the master branch.

The link step above is using g++ not gfortran, and I think that is the problem.

**Comment by drroe** [comment]
You have to force cpptraj to link in fortran libraries because it doesn't by default. The stand-alone configure for cpptraj determines during configure whether fortran linking is necessary. However, inside AmberTools the checks are skipped, and cpptraj is simply forced to use fortran linking (because it is assumed we are using arpack from AmberTools) via passing cpptraj's configure `--requires-flink`. My guess is that CMAKE isn't doing the equivalent thing.

**Comment by multiplemonomials** [comment]
Hmmm, CMake is supposed to be able to figure this stuff out by itself, it's never really been an issue before.  I'll take a look.

**Comment by multiplemonomials** [comment]
Hmm, I just tested RC5 on an Ubuntu install and it worked fine, apart from my previous fix to Miniconda installation not working (will commit a fix soon).

I think what might be happening is it is picking up a different version of Arpack from a different Amber build.  Can you try passing `-DFORCE_INTERNAL_LIBS=arpack` to CMake?

**Comment by dacase** [comment]
OK- I finally got a Linux gnu compilation to build.  I used -DFORCE_INTERNAL_LIBS=arpack.  Also hade to truncate my PATH so that a miniconda installation from another Amber build was not visible to the build script.  (I've always had trouble if there is an existing miniconda installation in my PATH, even though CMake insists on using this if it finds it.)

FWIW, here is my full cmake invocation:

cmake /home/case/test/amber18 -DCMAKE_INSTALL_PREFIX=/home/case/test/amber_ci -DCOMPILER=GNU -DFORCE_INTERNAL_LIBS=arpack -DBUILD_GUI=FALSE

The source code is post-RC5, with the cmake directory having the recent patches Jamie posted to master.  Compiler is gcc 5.4.0.

Next step: see if I can have any more luck with MacOS.

**Comment by hainm** [comment]
Have you tried with "make install -j4". I got compilation error at cpptraj step (in master branch; so It might not be relevant).

**Comment by dacase** [comment]
Sort of got cmake with MacOS to work: everything compiles OK, but the initial build checks set BUILD_GUI to OFF, so it never tries to compile xleap.  I couldn't find anything in the build report to indicate why the GUI part was turned off, but I probably don't know what to look for....

Is anyone else seeing this: i.e. xleap is not made using CMake on MacOS?

**Comment by multiplemonomials** [comment]
@dacase What happens if you set `-DBUILD_GUI=TRUE`?  Also, can you post a bug report so I can look at the cache file?

**Comment by multiplemonomials** [comment]
Also @hainm Can you please post a bug report?

**Comment by dacase** [comment]
1. Don't know how to post a bug report (unless you mean Amber's bugzilla forum?).  Are there particular files you need?
2. I'll try adding -DBUILD_GUI=TRUE when I get some free time.

**Comment by multiplemonomials** [comment]
Yeah @dacase there is a bug report script on the wiki:  http://ambermd.org/pmwiki/pmwiki.php/Main/CMake-Reporting-A-Bug

**Comment by dacase** [comment]
OK: the release is out, so I changed the title of this thread.  I installed with cmake on Linux, gnu 5.4, and ran the serial tests, all of which pass with the legacy build for the release code.  I now see to non-passing tests:

1.  the phmd test in Amber: non-roundoff errors
2. the Run.gbsa3 test in test/gbsa_xfin:  surface area energy is zero in current output, non-zero in saved output

Note: above is in master, not amber18-with-patches.  So I need to track down if these diffs are related to cmake, or to release vs. master branch diffs.  Just posting the comment here so it won't get lost.

**Comment by dacase** [comment]
Another minor glitch: on MacOS, with perl-5.26 installed from macports, I get this error:

cannot find inc::Module::Install in @INC   [exact wording was lost, but it is close to this]

I could install p5.26-app-cpanminus from macports, then run 
cpanm-5.26 inc::Module::Install.  After which the build ran fine.

However:  1. we can't expect users to do this
                  2. my solution writes outside of the install/build directories, and has the possibility
                        of messing up other perl-related software on the users' computer

Possible solution: if the perl configuration fails, just skip building mmpbsa.pl and FEW, rather than aborting the entire Amber build.  [Don't know how easy/hard this is to do.]

Maybe cmake knows how to get better solutions?  This is really a problem for the Dusseldorf people to address, so the "solution" above (skipping installation of mmpbsa.pl and FEW) might kind of force their hand.

**Comment by multiplemonomials** [comment]
OK yeah that should be pretty easy to do.  I already found a CMake script that checks if a Perl package is installed, that's how I check if I even need to build PerlMol.  I can just check if Module::Install is installed and if not, skip building PerlMol (which automatically disables FEW and mm_pbsa).

**Comment by multiplemonomials** [comment]
Actually wait no, that isn't quite right.  Module::Install is supposed to actually bundle itself within the source distribution, to prevent exactly this problem.  From the download page:
```
Rather than try to target one specific installer and/or make you add twisty workaround expressions to every piece of install code you write, Module::Install will copy part of itself into each module distribution it creates.

This allows new improvements to be used in your installers regardless of the age of the system a distribution is being installed on, at the cost of a small increase in the size of your distribution.
```

So,I'm not sure how this happened, and it seems like it shouldn't have been able to break like this.  I know it's annoying, but do you mind uninstalling Module::Install, and sending the full error message?
